File: example projects/example-r/validator.R
library(digest)

validate_email <- function(email) {
  if (is.null(email) || nchar(email) == 0) {
    return(FALSE)
  }
  
  email_regex <- "^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}$"
  return(grepl(email_regex, email))
}

hash_password <- function(password) {
  if (nchar(password) < 8) {
    stop("Password must be at least 8 characters")
  }
  
  return(digest(password, algo = "sha256"))
}

verify_password <- function(password, hash) {
  tryCatch({
    computed_hash <- hash_password(password)
    return(computed_hash == hash)
  }, error = function(e) {
    return(FALSE)
  })
}

sanitize_string <- function(input) {
  if (is.null(input) || nchar(input) == 0) {
    return("")
  }
  
  sanitized <- trimws(input)
  sanitized <- gsub("<", "&lt;", sanitized)
  sanitized <- gsub(">", "&gt;", sanitized)
  sanitized <- gsub('"', "&quot;", sanitized)
  sanitized <- gsub("'", "&#x27;", sanitized)
  
  return(sanitized)
}

validate_username <- function(username) {
  if (is.null(username) || nchar(username) == 0) {
    return(FALSE)
  }
  
  if (nchar(username) < 3 || nchar(username) > 20) {
    return(FALSE)
  }
  
  username_regex <- "^[a-zA-Z0-9_]+$"
  return(grepl(username_regex, username))
}

validate_password_strength <- function(password) {
  if (nchar(password) < 8) {
    return(FALSE)
  }
  
  has_uppercase <- grepl("[A-Z]", password)
  has_lowercase <- grepl("[a-z]", password)
  has_digit <- grepl("[0-9]", password)
  has_special <- grepl("[!@#$%^&*(),.?\":{}|<>]", password)
  
  criteria_count <- sum(has_uppercase, has_lowercase, has_digit, has_special)
  return(criteria_count >= 3)
}

check_password_complexity <- function(password) {
  results <- list(
    length_ok = nchar(password) >= 8,
    has_uppercase = grepl("[A-Z]", password),
    has_lowercase = grepl("[a-z]", password),
    has_digit = grepl("[0-9]", password),
    has_special = grepl("[!@#$%^&*(),.?\":{}|<>]", password)
  )
  
  results$score <- sum(unlist(results))
  results$is_strong <- results$score >= 4
  
  return(results)
}

File: example projects/example-swift/Validator.swift
import Foundation
import CryptoKit

class Validator {
    private static let emailRegex = try! NSRegularExpression(
        pattern: "^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}$"
    )
    
    private static let usernameRegex = try! NSRegularExpression(
        pattern: "^[a-zA-Z0-9_]+$"
    )
    
    static func validateEmail(_ email: String) -> Bool {
        guard !email.isEmpty else {
            return false
        }
        
        let range = NSRange(location: 0, length: email.utf16.count)
        return emailRegex.firstMatch(in: email, options: [], range: range) != nil
    }
    
    static func hashPassword(_ password: String) throws -> String {
        guard password.count >= 8 else {
            throw ValidationError.passwordTooShort
        }
        
        let data = Data(password.utf8)
        let hash = SHA256.hash(data: data)
        return hash.compactMap { String(format: "%02x", $0) }.joined()
    }
    
    static func verifyPassword(_ password: String, hash: String) -> Bool {
        do {
            let computedHash = try hashPassword(password)
            return computedHash == hash
        } catch {
            return false
        }
    }
    
    static func sanitizeString(_ input: String) -> String {
        guard !input.isEmpty else {
            return ""
        }
        
        return input.trimmingCharacters(in: .whitespacesAndNewlines)
            .replacingOccurrences(of: "<", with: "&lt;")
            .replacingOccurrences(of: ">", with: "&gt;")
            .replacingOccurrences(of: "\"", with: "&quot;")
            .replacingOccurrences(of: "'", with: "&#x27;")
    }
    
    static func validateUsername(_ username: String) -> Bool {
        guard !username.isEmpty else {
            return false
        }
        
        guard username.count >= 3 && username.count <= 20 else {
            return false
        }
        
        let range = NSRange(location: 0, length: username.utf16.count)
        return usernameRegex.firstMatch(in: username, options: [], range: range) != nil
    }
    
    static func validatePasswordStrength(_ password: String) -> Bool {
        guard password.count >= 8 else {
            return false
        }
        
        let hasUppercase = password.rangeOfCharacter(from: .uppercaseLetters) != nil
        let hasLowercase = password.rangeOfCharacter(from: .lowercaseLetters) != nil
        let hasDigit = password.rangeOfCharacter(from: .decimalDigits) != nil
        let hasSpecial = password.rangeOfCharacter(from: CharacterSet.alphanumerics.inverted) != nil
        
        let criteriaCount = [hasUppercase, hasLowercase, hasDigit, hasSpecial].filter { $0 }.count
        return criteriaCount >= 3
    }
}

enum ValidationError: Error {
    case passwordTooShort
    case invalidEmail
    case invalidUsername
}
